src/crypto/vss.rs

```diff
@@ -0,0 +1,270 @@
+use crate::errors::Error;
+use bitcoin::consensus::encode::{self, *};
+use bitcoin::PublicKey;
+use curv::arithmetic::traits::Converter;
+use curv::elliptic::curves::traits::*;
+use curv::{BigInt, FE, GE};
+use std::fmt;
+use std::io;
+use std::str::FromStr;
+
+// | name                 | size      | explaination                                                                                      |
+// | -------------------- | --------- | ------------------------------------------------------------------------------------------------- |
+// | sender_public_key    | 33        | indicates the signer who sends the vss                                                            |
+// | receiver_public_key  | 33        | indicates the signer to be received the vss                                                       |
+// | positive commitments | 64 \* len | commitments for secret value for r . an array of the points on the elliptic curve secp256k1.      |
+// | positive secret      | 32        | secret value for r to perform secret sharing scheme                                               |
+// | negative commitments | 64 \* len | commitments for secret value for (n - r). an array of the points on the elliptic curve secp256k1. |
+// | negative secret      | 32        | secret value for (n - r) to perform secret sharing scheme                                         |
+pub struct Vss {
+    pub sender_public_key: PublicKey,
+    pub receiver_public_key: PublicKey,
+    pub positive_commitments: Vec<Commitment>,
+    pub positive_secret: FE,
+    pub negative_commitments: Vec<Commitment>,
+    pub negative_secret: FE,
+}
+
+impl Vss {
+    fn new(sender_public_key: PublicKey,
+        receiver_public_key: PublicKey,
+        positive_commitments: Vec<Commitment>,
+        positive_secret: FE,
+        negative_commitments: Vec<Commitment>,
+        negative_secret: FE,) -> Self {
+            assert_eq!(positive_commitments.len(), negative_commitments.len());
+            Vss {
+                sender_public_key: sender_public_key,
+                receiver_public_key: receiver_public_key,
+                positive_commitments: positive_commitments,
+                positive_secret: positive_secret,
+                negative_commitments: negative_commitments,
+                negative_secret: negative_secret,
+            }
+        }
+}
+
+#[derive(Clone)]
+pub struct Commitment {
```

Comment:
Is there any reason to create this struct, not to use GE directly? 

Creating this struct is no problem. It's just my curiosity.

Reply:
To define method for encoding/decoding commitment.
We cant implement the method consensus_decode/consensus_encode on GE directly, right?

Reply:
Yes, it is right. But encoding/decoding for a commitment can be included in VSS encoding/decoding function.